### PR TITLE
Ensure whitespaces after semicolon in structural dir

### LINKED ISSUE #330: Ensure whitespaces after semicolon in structural dir
This rule should extend `angular-whitespace`.

Source https://github.com/mgechev/codelyzer/issues/314

### PR DESCRIPTION
**Pull request by wKoza**
closes #330

### CODE REVIEW

File: src/angularWhitespaceRule.ts [hunk 1]
```diff
@@ -12,6 +12,8 @@ const InterpolationClose = Config.interpolation[1];
 const InterpolationNoWhitespaceRe =new RegExp(`${InterpolationOpen}\\S(.*?)\\S${InterpolationClose}|${InterpolationOpen}\\s(.*?)\\S${InterpolationClose}|${InterpolationOpen}\\S(.*?)\\s${InterpolationClose}`);
 const InterpolationExtraWhitespaceRe =
   new RegExp(`${InterpolationOpen}\\s\\s(.*?)\\s${InterpolationClose}|${InterpolationOpen}\\s(.*?)\\s\\s${InterpolationClose}`);
+const SemicolonNoWhitespaceRe =new RegExp(/;\S(.*)/);
```
Comment (mgechev) [inline src/angularWhitespaceRule.ts]: Whitespace after `=`.

File: src/angularWhitespaceRule.ts [hunk 2]
```diff
@@ -26,7 +28,16 @@ const getReplacements = (text: ast.BoundTextAst, absolutePosition: number) => {
   ];
 };
 
-type Option = 'check-interpolation' | 'check-pipe';
+
+const getSemicolonReplacements = (text: ast.BoundDirectivePropertyAst, absolutePosition: number) => {
+
+  return [
+    new Lint.Replacement(absolutePosition, 1, "; ")
```
Comment (mgechev) [inline src/angularWhitespaceRule.ts]: Lets use single quotes.

File: src/angularWhitespaceRule.ts [hunk 3]
```diff
@@ -65,9 +76,42 @@ class InterpolationWhitespaceVisitor extends BasicTemplateAstVisitor implements
...
+      let expr = directive.split("=")[1].trim();
+      expr = expr.substring(1,expr.length-1).trim();
+
+      // Note that will not be reliable for different interpolation symbols
+      let error = null;
+
+      if (SemicolonNoWhitespaceRe.test(expr)) {
+        error = 'Missing whitespace after semicolon; expecting \'; expr\'';
+        const internalStart = expr.search(SemicolonNoWhitespaceRe)+1;
+        const start = prop.sourceSpan.start.offset + internalStart + directive.length - directive.split("=")[1].trim().length +1;
```
Comment (mgechev) [inline src/angularWhitespaceRule.ts]: Lets use single quotes and add space after the `+`.

File: src/angularWhitespaceRule.ts [hunk 4]
```diff
@@ -65,9 +76,42 @@ class InterpolationWhitespaceVisitor extends BasicTemplateAstVisitor implements
...
+        const absolutePosition = context.getSourcePosition(start-1);
```
Comment (mgechev) [inline src/angularWhitespaceRule.ts]: Whitespace around `-`.

**Comment by mgechev** [comment]
@wKoza awesome job man! I left some very minor style-related comments. Will provide further feedback after AngularUP.

**Comment by rokerkony** [comment]
hey @wKoza, nice job... I found one case which should be still probably taken in account:

template:
`<div *ngIf="codelyzer | number:'3;1'; fails"></div>`

The arguments of the pipe should not been checked as it could influence an output of that pipe.

Other question is if you shouldn't check for whitespace specifically. Now it is possible to have `awesome<TAB>|<TAB>pipe`

**Comment by mgechev** [comment]
I suppose we can generalize that we should not check string literals.

File: src/angularWhitespaceRule.ts [hunk 5]
```diff
@@ -65,9 +76,42 @@ class InterpolationWhitespaceVisitor extends BasicTemplateAstVisitor implements
...
+      const directive = (<any>prop.sourceSpan).toString();
+      let expr = directive.split("=")[1].trim();
+      expr = expr.substring(1,expr.length-1).trim();
+
+      // Note that will not be reliable for different interpolation symbols
+      let error = null;
+
+      if (SemicolonNoWhitespaceRe.test(expr)) {
+        error = 'Missing whitespace after semicolon; expecting \'; expr\'';
+        const internalStart = expr.search(SemicolonNoWhitespaceRe)+1;
```
Comment (rokerkony) [inline src/angularWhitespaceRule.ts]: what about adding 
```json

    "whitespace": [
      true,
      "check-branch",
      "check-decl",
      "check-module",
      "check-operator",
      "check-preblock",
      "check-separator",
      "check-type"
    ],
```

to tslint.json and not have a necessity to deal with this in CR? :)

btw. tslint is outdated and some rules are already gone. If you want, I can prepare some PR with proposal.

File: src/angularWhitespaceRule.ts (hunk 1 above)
Comment (wKoza) [inline src/angularWhitespaceRule.ts]: done

File: src/angularWhitespaceRule.ts (hunk 2 above)
Comment (wKoza) [inline src/angularWhitespaceRule.ts]: done

File: src/angularWhitespaceRule.ts (hunk 3 above)
Comment (wKoza) [inline src/angularWhitespaceRule.ts]: done

File: src/angularWhitespaceRule.ts (hunk 4 above)
Comment (wKoza) [inline src/angularWhitespaceRule.ts]: done

File: src/angularWhitespaceRule.ts [hunk 6]
```diff
@@ -65,9 +76,42 @@ class InterpolationWhitespaceVisitor extends BasicTemplateAstVisitor implements
   }
 }
 
+class SemicolonTempalteVisitor extends BasicTemplateAstVisitor implements ConfigurableVisitor {
+
+  visitDirectiveProperty(prop: ast.BoundDirectivePropertyAst, context: BasicTemplateAstVisitor): any {
+
+    if (prop.sourceSpan) {
+      const directive = (<any>prop.sourceSpan).toString();
+      let expr = directive.split("=")[1].trim();
```
Comment (wKoza) [inline src/angularWhitespaceRule.ts]: done

File: src/angularWhitespaceRule.ts [hunk 7]
```diff
@@ -65,9 +76,42 @@ class InterpolationWhitespaceVisitor extends BasicTemplateAstVisitor implements
...
+      expr = expr.substring(1,expr.length-1).trim();
```
Comment (wKoza) [inline src/angularWhitespaceRule.ts]: done

File: src/angularWhitespaceRule.ts (hunk 5 above)
Comment (wKoza) [inline src/angularWhitespaceRule.ts]: done.

Comment (mgechev) [inline src/angularWhitespaceRule.ts]: @rokerkony yes, it's best if we add the checks as part of CI.

A PR will be good.

**Pull request merged**